fixed HA CB and HA Gateway Templates

Alan-AcutePath opened the pull request.

msftclas commented:
@__Alan-AcutePath__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@Alan-AcutePath__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

jagilber commented:
tested PR for ha-broker deployment.
- tested with empty 'sqlServer' and deployment failed. is this new behavior where sqlServer needs to be specified for Azure SQL and on prem?
- tested with 'sqlServer' populated with Azure Sql server name. shows deployment succeeded, but it failed. rds deployment doesn't show it as ha. analyzing...
`//----------------------------------------------------------------------------------------
//2017-05-14T11:15:50.5777400-04:00   DEPLOYMENT: rdsdepha4   Microsoft.Template   Succeeded
{
		
    DeploymentName:  Microsoft.Template
    CorrelationId:  67bacc4d-21d9-4037-9ae1-ef9e13e80b29
    ResourceGroupName:  rdsdepha4
    ProvisioningState:  Succeeded
    Timestamp:  \/Date(1494774950577)\/
    Mode:  Incremental
    TemplateLink:  Microsoft.Azure.Management.ResourceManager.Models.TemplateLink
    TemplateLinkString:  
		Uri            : https://raw.githubusercontent.com/Alan-AcutePath/RDS-Templates/master/rds-deployment-ha-broker/azuredeploy.json
		ContentVersion : 1.0.0.0
		
    DeploymentDebugLogLevel:  RequestContent
		 ResponseContent
    Parameters:  {
		
                       existingAdminUsername:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       windowsServerVersion:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       sqlServer:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingVnetName:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       dnsServer:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingSubnetName:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingDomainName:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       primaryDbConnectionString:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingAdminPassword:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingBrokerHostname:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       existingBrokerAvailabilitySet:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                       clientAccessName:  Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels.DeploymentVariable
                   
		}
    ParametersString:  
		Name             Type                       Value     
		===============  =========================  ==========
		clientAccessName  String                     HARDCB    
		dnsServer        String                     dc        
		existingAdminPassword  SecureString                         
		existingAdminUsername  String                     vmadmin   
		existingBrokerAvailabilitySet  String                     cb-availabilityset
		existingBrokerHostname  String                     broker    
		existingDomainName  String                     rdsdepha4.lab
		existingSubnetName  String                     ADStaticSubnetrdsdepha4
		existingVnetName  String                     ADVNETrdsdepha4
		primaryDbConnectionString  String                     Server=tcp:sql-server-1538149580.database.windows.net
		1433;Initial Catalog=RdsCb;Persist Security Info=False;User ID=sql-administrator;Password=notMyP@ssw0rd;MultipleActiveResultSets=False;Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;
		sqlServer        String                     sql-server-1538149580
		windowsServerVersion  String                     2016-Datacenter
		
    Outputs:  null
    OutputsString:  null

		},
`

jagilber commented:
testing ha config manually on same deployment for azure sql. part of issue was my connection string.
this string worked manually:
DRIVER=SQL Server Native Client 11.0;Server=tcp:sql-server-1538149580.database.windows.net,1433;Database=RdsCb;Uid=sql-administrator@sql-server-1538149580;Pwd=notMyP@ssw0rd;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;

redeploying

mmarch commented on rds-deployment-ha-broker/README.md:
```diff
@@ -35,7 +35,7 @@ This template expects the same names of resources from RDS deployment, if resour
 Click the button below to deploy
 
-<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2FRDS-templates%2Fmaster%2Frds-deployment-ha-broker%2Fazuredeploy.json" target="_blank">
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAlan-AcutePath%2FRDS-Templates%2Fmaster%2Frds-deployment-ha-broker%2Fazuredeploy.json" target="_blank">
```
please revert this, buttons should point to master

mmarch replied:
must point at Azure/RDS-Templates

mmarch commented on rds-deployment-ha-broker/azuredeploy.json:
```diff
@@ -102,11 +102,11 @@
     "scriptParameters": "[concat(variables('p').broker, variables('p').connstr, variables('p').domain, variables('p').user, variables('p').pwd, variables('p').can, variables('p').sqlnclient, variables('p').dnsserver, variables('p').sqlserver)]",
     "assetLocation": "https://raw.githubusercontent.com/Azure/azure-QuickStart-Templates/master/rds-deployment",
-    "scriptLocation": "https://rdsonazureiaas9114.blob.core.windows.net/msftrdsdeployment",
+    "scriptLocation": "https://raw.githubusercontent.com/Alan-AcutePath/RDS-Templates/master/rds-deployment-ha-broker/Scripts",
```
must point at Azure/RDS-Templates

mmarch commented on .vs/ProjectSettings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
please remove the 4 files from .vs subfolder from this pr

mmarch commented on rds-deployment-ha-gateway/README.md:
```diff
@@ -37,7 +37,7 @@ This template expects the same names of resources from RDS deployment, if resour
 Click the button below to deploy
 
-<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Frds-deployment-HA%2Fazuredeploy.json" target="_blank">
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAlan-AcutePath%2FRDS-Templates%2Fmaster%2Frds-deployment-ha-gateway%2Fazuredeploy.json" target="_blank">
```
must point at Azure/RDS-Templates

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -248,9 +166,7 @@
-      "dependsOn": [
-        "Microsoft.Network/loadBalancers/loadBalancer"
-      ],
+      "dependsOn": [],
```
remove "dependsOn" completely

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -8,16 +8,23 @@
       }
     },
-    "loadBalancer": {
+    "gatewayloadBalancer": {
```
why can't we keep "loadBalancer" as the parameter name? remind me, is there another load balancer in the deployment?

if you're chose to keep it as "gatewayloadBalancer" follow camelcase: "gatewayLoadBalancer"

Alan-AcutePath replied:
yes, there is another load balancer for ad deployed by the basic template. Just like to make it more clear. I will update it to follow the naming convention.

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -8,16 +8,23 @@
       },
       "defaultValue": "loadBalancer"
     },
+    "gatewaypublicip": {
```
camelcase: "gatewayPublicIp"

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -8,16 +8,23 @@
     "backendAddressPools": {
       "type": "string",
-      "defaultValue": "LBBAP",
+      "defaultValue": "LBBE",
```
is this right?

if yes, update description below as well

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -367,7 +283,7 @@
             "autoUpgradeMinorVersion": true,
             "settings": {
-              "fileUris": [ "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/rds-deployment-ha-gateway/Scripts/WebAndGwFarmAdd_PostConfig1.1.ps1" ]
+              "fileUris": [ "https://raw.githubusercontent.com/Alan-AcutePath/RDS-Templates/master/rds-deployment-ha-gateway/Scripts/WebAndGwFarmAdd_PostConfig1.1.ps1" ]
```
must point at Azure/RDS-Templates

suggestion: use assetLocation parameter, and use concat() function

Alan-AcutePath commented:
The HA gateway template doesn't need to make any changes to the existing load balancer deployed by  basic template. All it needs is to attach the gateway nic to backend address pool.

Alan-AcutePath commented:
remove the vs files, updated the links to azure/rds-templates, and parameter names to follow the naming conventions

mmarch commented:
@Alan-AcutePath,

Gang, can you reply to Jason's comment
* tested with empty 'sqlServer' and deployment failed. is this new behavior where sqlServer needs to be specified for Azure SQL and on prem?

What is the scenario where sqlServer parameter is empty string, can you clarify?
In other words, are these changes in rds-deployment-ha-broker/Scripts/RdcbHaPostConfig.ps1  really required?

mmarch commented on rds-deployment-ha-broker/Scripts/RdcbHaPostConfig.ps1:
```diff
@@ -20,7 +20,7 @@ param
      [string]$sqlClientUrl,
      [string]$dnsServer,
-     [string]$sqlServer
+     [string]$sqlServer=""
```
are changes to the script here really required?

what is the scenario where sqlServer is null/empty string?

Alan-AcutePath replied:
just set it default as empty string. But it didn't fix the issue when the parameter is not populated. What I did is the variable section in the template to put quote around the parameter even it is empty, the parameter will still pass in to prevent the issue. it works for me

mmarch commented on rds-deployment-ha-gateway/azuredeploy.json:
```diff
@@ -8,13 +8,20 @@
       },
       "defaultValue": "loadBalancer"
     },
+    "gatewayPublicip": {
```
nitpicking here, 
but want to see "Ip" starting with capital too, i.e. "gatewayPublicIp"

Alan-AcutePath replied:
sure. I will update it.

Alan-AcutePath commented:
the sqlserver parameter should not be required. To fix this, I added the quote around the parameter so it should be passed into the script even it is null/empty string. And I added a switch in the script to prevent doing computer group related configuration if sqlServer is not populated. It works for me.

Alan-AcutePath commented:
removed the vs files
